Story#76: npm module that can be used for `fs` related functions as Promises

sanakanchha opened the pull request:
- a npm library that can be used for file system interaction 
- added debug logging
- added unit tests

Commit:
create a fs library, add tests plus fix lint errors

Commit:
add debug logging to the library

brozeph commented on .eslintrc.yml:
```diff
@@ -3,41 +3,30 @@
     modules: true
   env:
     browser: false
-    es6: true
```
Recommend adding `es6` and `mocha` back in. You'll also want to add the following:

```yml
globals:
  should: true
```

And then just above `no-undefined: 2`, add `no-undef: 2`

sanakanchha replied:
this has been taken care of

sanakanchha replied:
this has been taken care of

brozeph commented on src/index.js:
```diff
@@ -1,12 +1,14 @@
-import debug from 'debug';
+import debugLog from 'debug';
 import fs from 'graceful-fs';
 import mkdirp from 'mkdirp';
```
Is this really a run-time dependency for fs-wrapper? We may want to consider a couple of options here - one is to remove this dependency all-together (i.e. take out the methods in the module that utilize it or write our own version of it in this module). If that does not seem tenable, then this dependency needs to be added to the `dependencies` list in the `package.json` file.

cmerchant replied:
it looks to be a logging thing. I see in the code below, debug('examining path %s', directoryPath);

Do we want to use the winston logger like we do in the other projects? How do other open source projects log things generally?

brozeph replied:
For logging within generic Node modules, [`debug`](https://www.npmjs.com/package/debug) is pretty much the defacto. The problem with forcing something like 'Winston` or `Bunyan` is that you're now creating a dependency on the Open Source consumer of this module (which could be anyone in the world) to use the logging paradigm we've chosen for our Linux player.

cmerchant replied:
got it, so it sounds like maybe we should add debug to the packages.json file unless there is a reason not to do that

sanakanchha replied:
the method 'ensurePath' uses mkdirp and is being used across all projects. I would prefer to add it as a dependency to the package.json.

cmerchant replied:
I thought we were talking about the debugLog dependency, not mkdirp. They both need to be in packages

brozeph commented on test/common.js:
```diff
@@ -0,0 +1,147 @@
+	TEST_CACHE_PATH = 'test/sandbox/test-cache',
+	TEST_CLIENT_ID = '982f0d7f23a94811849f703ef8150401',
+	TEST_TEMP_PATH = 'test/sandbox/test-temp',
+	TEST_SHARED_SECRET = '68JLP0L6M10/kfvtXNQCrbXe3nz8g1smyCjTblcy4OKbd4cnxf0anuxvLZ0T4hZw8hMi8w8pPvezd6AcBwa8Hw==';
```
## This is a grave mistake.

Functioning PlayNetwork API credentials have been publicly exposed on the Internet as a result of this pull request.

This issue must be considered with extreme seriousness - for the time from when this commit was pushed to GitHub until the secret for the clientId referenced in this PR was changed, there was a huge amount of risk present to PlayNetwork made open and public to the Internet. While we have corrected this on the backend and we will chalk this up to a learning experience this time around, I've worked in companies where people would have been let go for this type of breach because it borders on gross negligence. 

Any code submitted to a public repository requires an extra layer of in-depth checks in the code *WELL IN ADVANCE* of performing even a local commit occurs - this fact was spelled out clearly in the story item and discussed verbally prior to the commencement of this work. Even if one were to have committed this locally and then removed it on a subsequent commit, the history of these commits and the changes are published to GitHub and, in this case, these credentials still would have been exposed.

brozeph replied:
It goes without saying - remove these credentials and everything else that is PlayNetwork proprietary or that references the `music-nodejs` app (presumably where these tests were copied from) entirely.

brozeph reviewed (changes requested):
This is an absolute non-starter.

Commit:
remove common file

cmerchant commented on test/mocha.opts:
```diff
@@ -0,0 +1,5 @@
+--require babel-register
+--require babel-polyfill
+--require test/common
```
remove test/common if it is no needed here. You don't have common checked in.

sanakanchha replied:
I have it checked in now with just a few helper methods

cmerchant commented on test/unit/src/index.js:
```diff
@@ -0,0 +1,238 @@
+import fs from 'graceful-fs';
```
You will have to define the constants that were in common to get these tests to work.

cmerchant replied:
there are also some methods here that were defined in common

cmerchant replied:
Tests are passing now. Thanks!

Commit:
address lint rule comment

Commit:
suppress lint errors

Commit:
add common helper module

sanakanchha commented:
I have addressed all the comments. Also tested this module works with music-nodejs as well as settings-nodes using 'npm link'. Will be creating PR for these apps with the changes accordingly.

Commit:
add dependencies to be able to run tests

brozeph commented:
Missing dependency:

```
$ npm test

> fs-promise-util@1.0.0 test /Volumes/Macintosh HD 2/Projects/PlayNetwork/fs-promise-util
> gulp lint && nyc -- _mocha ./test/unit

[10:03:08] Requiring external module babel-register
[10:03:08] Using gulpfile /Volumes/Macintosh HD 2/Projects/PlayNetwork/fs-promise-util/gulpfile.babel.js
[10:03:08] Starting 'lint'...
[10:03:09] Finished 'lint' after 1.06 s
module.js:471
    throw err;
    ^

Error: Cannot find module 'chai-as-promised'
```

brozeph commented:
Missing documentation: no methods or usage is discussed in the readme.md

brozeph commented:
Missing dependency:

```
$ npm test

> fs-promise-util@1.0.0 test /Volumes/Macintosh HD 2/Projects/PlayNetwork/fs-promise-util
> gulp lint && nyc -- _mocha ./test/unit

[10:05:48] Requiring external module babel-register
[10:05:48] Using gulpfile /Volumes/Macintosh HD 2/Projects/PlayNetwork/fs-promise-util/gulpfile.babel.js
[10:05:48] Starting 'lint'...
[10:05:49] Finished 'lint' after 699 ms
module.js:471
    throw err;
    ^

Error: Cannot find module 'sinon'
```

brozeph commented:
Code coverage report broken (see 0% at bottom of output):

```
$ npm test

> fs-promise-util@1.0.0 test /Volumes/Macintosh HD 2/Projects/PlayNetwork/fs-promise-util
> gulp lint && nyc -- _mocha ./test/unit

[10:06:39] Requiring external module babel-register
[10:06:39] Using gulpfile /Volumes/Macintosh HD 2/Projects/PlayNetwork/fs-promise-util/gulpfile.babel.js
[10:06:39] Starting 'lint'...
[10:06:40] Finished 'lint' after 728 ms


  src/index
    #appendFile
      ✓ should reject with invalid path
    #ensurePath
      ✓ should reject given undefined path
      ✓ should reject given empty path
      ✓ should reject given cache path already exists but is not a folder
      ✓ should resolve given cache path already exists as a directory
      ✓ should resolve given cache path already exists as symbolic link
      ✓ should resolve given valid path
    #exists
      ✓ should resolve to false given undefined filePath
      ✓ should resolve to false given unknown filePath
      ✓ should resolve to true given existing file
    #prune
      ✓ should reject given undefined directoryPath
      ✓ should reject given empty directoryPath
      ✓ should reject given undefined retainCount
      ✓ should reject given negative retainCount
      ✓ should reject given invalid directoryPath
      ✓ should resolve given empty TEST_CACHE_PATH
      ✓ should resolve and remove file given TEST_CACHE_PATH with file
      ✓ should resolve, remove, and retain file given TEST_CACHE_PATH with files
    #readAndSort
      ✓ should reject given undefined directoryPath
      ✓ should reject given empty directoryPath
      ✓ should reject given undefined retainCount
      ✓ should reject given negative retainCount
      ✓ should reject given invalid directoryPath
      ✓ should resolve given empty TEST_CACHE_PATH
      ✓ should resolve and remove file given TEST_CACHE_PATH with file
      ✓ should resolve, remove, and retain file given TEST_CACHE_PATH with files (56ms)
    #tryWriteFile
      ✓ should resolve given undefined filePath
      ✓ should resolve given and file written given filePath


  28 passing (425ms)


=============================== Coverage summary ===============================
Statements   : Unknown% ( 0/0 )
Branches     : Unknown% ( 0/0 )
Functions    : Unknown% ( 0/0 )
Lines        : Unknown% ( 0/0 )
================================================================================
```

Commit:
address cr comments: add dev-dependencies

Commit:
address cr comment: fix code coverage report

sanakanchha commented:
@brozeph all comments addressed except the documentation. Working on the readme.md file now.

brozeph commented:
OK, this is getting much closer!

```
=============================== Coverage summary ===============================
Statements   : 73.87% ( 82/111 )
Branches     : 68.52% ( 37/54 )
Functions    : 72.73% ( 24/33 )
Lines        : 76.19% ( 80/105 )
================================================================================
```

Let's get these coverage numbers up - ideally we target the 90s. We'll want to put a coveralls.io badge on this bad boy along with a travis.ci build status badge before we publish to npm to indicate that we've put in work to ensure it operates as expected and that we can maintain it moving forward with low risk to backwards incompatibility.

Commit:
add documentation

Commit:
documentation work

Commit:
style adjustments in readme.md

Commit:
add test plus example code in the documentation

Commit:
add tests plus examplify documentation

brozeph commented:
Need to add an item to the `.gitignore` file:

```
$ git status
On branch story#76
Your branch is up-to-date with 'origin/story#76'.
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	.nyc_output/

nothing added to commit but untracked files present (use "git add" to track)
```

Commit:
lstat: add test plus examplify documentation

Commit:
add nyc.output file to .gitignore

Commit:
add documentation for method prune

Commit:
add test for increased coverage

sanakanchha commented:
bumped up the coverage above 80%
============= Coverage summary ====================
Statements   : 81.08% ( 90/111 )
Branches     : 68.52% ( 37/54 )
Functions    : 81.82% ( 27/33 )
Lines        : 82.86% ( 87/105 )
================================================

Commit:
fix documentation

sanakanchha commented:
@brozeph should I talk to devops to setup the travis build?

brozeph commented:
@deepakdahal - can you go to npmjs.com and create an account? When done, please add your username to this thread...

sanakanchha commented:
@brozeph for some reason, I missed this comment. account created. username: sanakanchha
